### ISSUE TITLE
Import completion suggests items that will never work

### ISSUE THREAD

**Issue by cartermp** [dotnet/roslyn#37198]
**Version Used**: 

Microsoft Visual Studio Community 2019 Int Preview
Version 16.3.0 Preview 1.0 [29111.216.d16.3]
VisualStudio.16.IntPreview/16.3.0-pre.1.0+29111.216.d16.3
Microsoft .NET Framework
Version 4.8.03752

**Steps to Reproduce**:

Create a new .NET Core console app.

Write code like this:

```csharp
using System;

namespace ConsoleApp114
{
    class Foo<T> where T: //<<--- cursor here!

    class Program
    {
        static void Main(string[] args)
        {
            Console.WriteLine("Hello World!");
        }
    }
}

```

**Expected Behavior**:

Only generic constraints, or nothing at all.

**Actual Behavior**:

![image](https://user-images.githubusercontent.com/6309070/61160483-72498480-a4b4-11e9-928e-13ad992ad158.png)

**Comment by CyrusNajmabadi** [comment]
@genlu We have a bunch of filtering/checking in normal symbol completion.  are we just not executing it for import-completion for perf reasons?

**Comment by genlu** [comment]
We check to see if we are in the context for types. Not sure I understand the issue in the screenshot above (didn't have my coffee today so I might be missing something obvious here :-)): those seems fine since `where : <type>` is valid and the behavior above is consistent with  regular symbol completion?

**Comment by cartermp** [comment]
I suppose my issue here is that `where : <type>`  is too broad for filtering. Another example:

![image](https://user-images.githubusercontent.com/6309070/61243333-913a5780-a6fc-11e9-94cd-58f9993bd93f.png)

Generally, we should only be suggesting things that we know will compile.

Extended completion just exacerbates this since it places many more things in the completion window.

**Comment by genlu** [comment]
Ah I see. So you want only interface and non-sealed class types to be shown here? That makes sense.

**Comment by CyrusNajmabadi** [comment]
Interestingly, i thought we had that with the old native LS.  maybe we lost it? 

also, remember that types can have nested types.  so you can do this:  `where T : OuterSealedClass.InnerInterface`.

**Comment by genlu** [comment]
> also, remember that types can have nested types. so you can do this: where T : OuterSealedClass.InnerInterface.

Oh you are right, @jasonmalinowski mentioned this on another issue as well. I guess this is why we have to include all those types here. Maybe we can implement this additional filtering as part of the target completion?

**Comment by CyrusNajmabadi** [comment]
Well, my point was simpy this:  filtering is fine.  just don't *overfilter* :)

So we can filter out sealed types that don't contain compatible public nested types for example :)

**Comment by genlu** [comment]
> So we can filter out sealed types that don't contain compatible public nested types for example :)

Implement accurate filter for imported symbols might be fine, since the number of types are relatively small. So we should definitely explore that. But for import completion, we only have data for top level types, and it might be expensive to check nested type symbols for all of them (I might be wrong, the point is some test is required here).

With target completion, the idea is to filter out things can't be used at current location, so potentially it doesn't need to consider the nested types scenario. FYI @dpoeschl

**Comment by jasonmalinowski** [comment]
Target completion is going to filter away type names in general since a type name can't ever match a type -- it doesn't have a type _per se_. "Filter to types that have a method that target type" would be both really expensive and probably not useful in some cases.

I think @Pilchie said ages ago the C# IDE did try to filter to exception-derived types when in a catch, and the cost of doing so and handling the nested case exceeded any benefit and we just killed it.

**Comment by genlu** [comment]
I guess what I'm suggesting re: target completion is to expand it to cover type usage in inheritance and generic constraint scenario, such that say types can't be inherited will not be shown (I think I mentioned this to @dpoeschl before)

**Comment by dpoeschl** [comment]
> Which in my opinion has some similarities with filtering out objects that can't be converted to some target type.

It could be an aspect of a broader "targeted completion", if not "target-typed" exactly. I'm not sure how expensive this would be (like the exception-derived types thing Jason mentions).

**Comment by genlu** [comment]
> I'm not sure how expensive this would be
Yeah, the perf would definitely be an issue. What about making targeted completion to use the same model as expander? i.e. only trigger the computation when user select the targeted filter. We will need to discuss with editor team of course but it will enable us to do more expensive operations w/o worrying about typing perf regression.

**Comment by sharwell** [comment]
The proposal here is to only offer the following for completing a generic type constraint:

1. Interfaces
2. Non-sealed types
3. Type parameters
4. Sealed types with one or more nested types (optionally filtering this to only consider nested types meeting this same set of conditions)

**Comment by CyrusNajmabadi** [comment]
Closing out as this is working as expected now.
